### DISCUSSION TITLE
Need help with options that may or may not have a parameter value

### DISCUSSION THREAD

**Discussion by sean-neeley** [remkop/picocli#1951] [Q&A] [answered]
I have the following code that correctly outputs this usage message:
`Usage: test [-hV] [-y[=FILE]] USERNAME`

I would like to be able to run it one of two ways: with arguments '-y USERNAME' and '-y=FILE USERNAME'.  However '-y USERNAME' is treated as '-y=USERNAME' (which I do not want) and gives a syntax error, when instead I want it to be equivalent to '-y=blank USERNAME'.   Can I have '=' be the only command separator instead of both space and '='?  Or is there another solution to this?  Thanks.

```
import picocli.CommandLine;
import picocli.CommandLine.Command;
import picocli.CommandLine.Option;
import picocli.CommandLine.Parameters;

import java.util.concurrent.Callable;

@Command(name = "test", mixinStandardHelpOptions = true, version = "1.0", description = "Test program.")
class Test implements Callable<Integer> {

    @Option(names = { "-y" },
    		paramLabel = "FILE",
    		arity = "0..1",
    		fallbackValue = "blank", 
    		description = "Do task with optional input file.")
    private String keyExchangeFile;

    @Parameters(index = "0", 
    		description = "User name", 
    		paramLabel="USERNAME", 
    		arity = "1") 
    private String userName;
    
    @Override
    public Integer call() throws Exception { 
    	// your business logic goes here...
        return 0;
    }

    public static void main(String... args) {
        CommandLine cmd = new CommandLine(new Test());
    	cmd.setSeparator("=");
        int exitCode = cmd.execute(args);
        System.exit(exitCode);
    }
}
```

**Comment by remkop** [comment, marked as answer]
This is a common problem for options that have an optional value.
Please see this documentation on [Optional Parameter Limitations](https://picocli.info/#_optional_parameter_limitations) in the user manual.

Your best solution may be to use the [example `IParameterPreprocessor` from the user manual](https://picocli.info/#IParameterPreprocessor_Parser_Plugin) in your application.
That preprocessor implementation customizes the parser, such that `VALUE` in `--option=VALUE` is interpreted as the option parameter, and in `--option VALUE` (without the `=` separator), `VALUE` is interpreted as the positional parameter.
I believe this is exactly what you are looking for.

**Comment by sean-neeley** [comment]
Thank you, that is the exact issue I'm running into and that example is perfect.  Somehow I missed it in the documentation.  I appreciate the help!
